golem-worker-executor-base/src/durable_host/golem/mod.rs

```diff
@@ -27,12 +30,96 @@ use crate::get_oplog_entry;
 use crate::metrics::wasm::record_host_function_call;
 use crate::model::InterruptKind;
 use crate::preview2::golem;
-use crate::preview2::golem::api::host::{OplogIndex, PersistenceLevel, RetryPolicy};
+use crate::preview2::golem::api::host::{
+    HostGetWorkers, OplogIndex, PersistenceLevel, RetryPolicy,
+};
 use crate::workerctx::WorkerCtx;
 use golem_common::model::oplog::{OplogEntry, WrappedFunctionType};
 use golem_common::model::regions::OplogRegion;
 use golem_common::model::{PromiseId, TemplateId, WorkerId};
 
+#[async_trait]
+impl<Ctx: WorkerCtx> HostGetWorkers for DurableWorkerCtx<Ctx> {
+    async fn new(
+        &mut self,
+        template_id: golem::api::host::TemplateId,
+        filter: Option<golem::api::host::WorkerAnyFilter>,
+        precise: bool,
+    ) -> anyhow::Result<Resource<GetWorkersEntry>> {
+        record_host_function_call("golem::api::get-workers", "new");
+        let entry = GetWorkersEntry::new(template_id.into(), filter.map(|f| f.into()), precise);
+        let resource = self.as_wasi_view().table_mut().push(entry)?;
+        Ok(resource)
+    }
+
+    async fn get_next(
+        &mut self,
+        self_: Resource<GetWorkersEntry>,
+    ) -> anyhow::Result<Option<Vec<golem::api::host::WorkerMetadata>>> {
+        record_host_function_call("golem::api::get-workers", "get_next");
+        let (template_id, filter, count, precise, next_cursor) = self
+            .as_wasi_view()
+            .table()
+            .get::<GetWorkersEntry>(&self_)
+            .map(|e| {
+                (
+                    e.template_id.clone(),
+                    e.filter.clone(),
+                    e.count,
+                    e.precise,
+                    e.next_cursor.clone(),
+                )
+            })?;
+
+        let cursor = *next_cursor.read().unwrap();
+
+        if let Some(cursor) = cursor {
+            let (new_cursor, workers) = self
+                .state
+                .get_workers(&template_id, filter, cursor, count, precise)
+                .await?;
+
+            *next_cursor.write().unwrap() = new_cursor;
+
+            Ok(Some(workers.into_iter().map(|w| w.into()).collect()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn drop(&mut self, rep: Resource<GetWorkersEntry>) -> anyhow::Result<()> {
+        record_host_function_call("golem::api::get-workers", "drop");
+        self.as_wasi_view()
+            .table_mut()
+            .delete::<GetWorkersEntry>(rep)?;
+        Ok(())
+    }
+}
+
+pub struct GetWorkersEntry {
+    template_id: golem_common::model::TemplateId,
+    filter: Option<golem_common::model::WorkerFilter>,
+    precise: bool,
+    count: u64,
+    next_cursor: Arc<RwLock<Option<u64>>>,
```

Comment:
You don't have to protect it because the host functions never run in parallel. In fact the whole context (wasmtime's, which hold `DurableCtx` as its user-defined part) is protected by a big mutex.
So just use `Option<u64>` here, and use `get_mut` when getting it from the table instead of `get`.